Editorial edits on fast open (#544)
ietf-tapswg/api-drafts: draft-ietf-taps-impl.md

Several protocols allow sending higher-level protocol or application data within the first packet of their protocol establishment, such as TCP Fast Open {{!RFC7413}} and TLS 1.3 {{!RFC8446}}. This approach is referred to as sending Zero-RTT (0-RTT) data. This is a desirable property, but poses challenges to an implementation that uses racing during connection establishment.

> If the application has 0-RTT data to send in any protocol handshakes, it needs to provide this data before any handshake is initiated. When racing, this means that the data should be provided before the process of connection establishment has begun. If the application wants to send 0-RTT data, it indicates this to the implementation by setting the Idempotent flag in the message context to true and providing initial data to send at the connection initiation. In general, 0-RTT data may be replayed (for example, if a TCP SYN contains data, and the SYN is retransmitted, the data will be retransmitted as well but may be considered as a new connection instead of a retransmission). Also when racing connections, different leaf nodes may send the same data independently. If data is truly idempotent, this should be permissible.

Reviewer: ```suggestion
If the application has 0-RTT data to send in any protocol handshakes, it needs to provide this data before any handshake is initiated. When racing, this means that the data should be provided before the process of connection establishment has begun. If the application wants to send 0-RTT data, it indicates this to the implementation by setting the Idempotent flag in the message context to true and providing initial data to send at the connection initiation. In general, 0-RTT data may be replayed (for example, if a TCP SYN contains data, and the SYN is retransmitted, the data will be retransmitted as well but may be considered as a new connection instead of a retransmission). Also, when racing connections, different leaf nodes may send the same data independently. If data is truly idempotent, this should be permissible.
```

Once the application has provided its 0-RTT data, an implementation should keep a copy of this data and provide it to each new leaf node that is started and for which a 0-RTT protocol is being used.

It is also possible that protocol stacks within a particular leaf node use 0-RTT handshakes without any idempotent application data. For example, TCP Fast Open could use a Client Hello from TLS as its 0-RTT data, shortening the cumulative handshake time.

> 0-RTT handshakes often rely on previous state, such as TCP Fast Open cookies, previously established TLS tickets, or out-of-band distributed pre-shared keys (PSKs). Implementations should be aware of security concerns around using these tokens across multiple addresses or paths when racing. In the case of TLS, any given ticket or PSK should only be used on one leaf node as servers likely will not accept dublicate tickets to prevent replays (see section-8.1 {{?RFC8446}}). If implementations have multiple tickets available from a previous connection, each leaf node attempt can use a different ticket. In effect, each leaf node will send the same early application data, yet encoded (encrypted) differently on the wire.

Reviewer: ```suggestion
0-RTT handshakes often rely on previous state, such as TCP Fast Open cookies, previously established TLS tickets, or out-of-band distributed pre-shared keys (PSKs). Implementations should be aware of security concerns around using these tokens across multiple addresses or paths when racing. In the case of TLS, any given ticket or PSK should only be used on one leaf node, since servers will likely reject duplicate tickets in order to prevent replays (see section-8.1 {{?RFC8446}}). If implementations have multiple tickets available from a previous connection, each leaf node attempt can use a different ticket. In effect, each leaf node will send the same early application data, yet encoded (encrypted) differently on the wire.
```
